README.md
# vue3-cron

A Cron expression UI component based on Vue3 and Ant Design Vue.

English | [简体中文](./README-zh_CN.md)

## Preview

![](public/preview-en.png)

## Supported format

```
*    *    *    *    *    *    *
┬    ┬    ┬    ┬    ┬    ┬    ┬
│    │    │    │    │    |    └ year (currentYear - 2099)
│    │    │    │    │    └───── day of week (0 - 7) (0 or 7 is Sun)
│    │    │    │    └────────── month (1 - 12)
│    │    │    └─────────────── day of month (1 - 31)
│    │    └──────────────────── hour (0 - 23)
│    └───────────────────────── minute (0 - 59)
└────────────────────────────── second (0 - 59)
```

| Field  | Required | Value range      | Allowed wildcard |
|--------|----------|------------------|------------------|
| Second | Yes      | 0-59             | , - * /          |
| Minute | Yes      | 0-59             | , - * /          |
| Hour   | Yes      | 0-23             | , - * /          |
| Date   | Yes      | 1-31             | , - * / L W      |
| Month  | Yes      | 1-12             | , - * /          |
| Week   | Yes      | 0-7 or SUN-SAT   | , - * / L #      |
| Year   | No       | currentYear-2099 | , - * /          |

## Installation

```
npm install vue3-cron-antd --save
```

## Usage

### Import

```typescript
// Global import.
import { createApp } from 'vue';
import AntDesignVue from 'ant-design-vue';
import 'ant-design-vue/dist/antd.css';
import Vue3Cron from 'vue3-cron-antd';

import App from './App.vue';

const app = createApp(App);

app.use(AntDesignVue);
app.use(Vue3Cron);

app.mount('#app');
```
or
```vue
<script>
import Vue3Cron from 'vue3-cron-antd';
// Single import.
export default {
    components: {
        Vue3Cron,
    },
}
</script>
```

### Use

```vue
<template>
    <vue3-cron v-model="expression" :locale="locale" />
</template>

<script>
    export default {
        data() {
            return {
                expression: '* * * * * ?',
                locale: 'cn', // set 'cn' or 'en', default is 'cn'.
            };
        },
    }
</script>
```


CHANGELOG.md
# Change Log

## 1.1.0 (2022-10-31)


### Features





## 1.0.0 (2022-10-29)


### Features





vite.config.ts
import { fileURLToPath, URL } from 'node:url';
import { resolve } from 'path';
import { defineConfig } from 'vite';
import vue from '@vitejs/plugin-vue';
import vueJsx from '@vitejs/plugin-vue-jsx';
import cssInjectedByJsPlugin from 'vite-plugin-css-injected-by-js';

// https://vitejs.dev/config/
export default defineConfig({
    plugins: [vue(), vueJsx(), cssInjectedByJsPlugin(),],
    resolve: {
        alias: {
            '@': fileURLToPath(new URL('./src', import.meta.url)),
        },
    },
    build: {
        lib: {
            entry: resolve(__dirname, 'src/index.ts'),
            name: 'Vue3Cron',
            formats: ['es', 'cjs', 'umd', 'iife'],
        },
        rollupOptions: {
            // 确保外部化处理那些你不想打包进库的依赖
            external: ['vue', 'ant-design-vue', 'vue-i18n', 'cron-parser'],
            output: {
                // 在 UMD 构建模式下为这些外部化的依赖提供一个全局变量
                globals: {
                    'vue': 'Vue',
                    'ant-design-vue': 'antd',
                    'vue-i18n': 'VueI18n',
                },
            },
        },
    },
})


src/shared/utils.ts
import { WEEKS, WEEK, WEEK_INDEX_REGEXP, WEEK_LETTER_REGEXP } from './constants.ts';
import I18n from '@/i18n';

/**
 * 生成指定数组.
 *
 * @param min - 区间最小值.
 * @param max - 区间最大值.
 * @param labels - 显示名称.
 */
export function generateSpecifies(min: number, max: number, labels: Array<string>): Array<{ value: number, label: string }> {
    const specifies = [];

    let index = 0;

    for (let specify = min; specify <= max; specify += 1) {
        specifies.push({ value: specify, label: labels ? labels[index] : specify.toString() });
        index += 1;
    }

    return specifies;
}

/**
 * 日期时间字段补零.
 *
 * @param value - 待补零处理的数字.
 */
export function zerofill(value: number): string {
    const prefix = (value < 10) ? '0' : '';

    return `${prefix}${value}`;
}

/**
 * 转换星期数字为缩写.
 *
 * @param value - .
 */
export function weekNumberToLetter(value: string): string {
    return value.replace(WEEK_INDEX_REGEXP, value => WEEKS.find(({ index }) => [index].includes(value))?.abbr || value);
}

/**
 * 转换星期缩写为数字.
 *
 * @param value - .
 */
export function weekLetterToNumber(value: string): string {
    return value.replace(WEEK_LETTER_REGEXP, value => WEEKS.find(({ abbr }) => abbr === value)?.index || value);
}

/**
 * 星期字段的 <a-input-number> 组件 formatter 函数.
 *
 * @param value - .
 * @param locale - 国际化字段.
 */
export function formatterWeek(value: string, locale: string): string {
    const day = WEEKS.find(({ index }) => index === value)?.value;

    return I18n[locale][WEEK][day];
}

/**
 * 星期字段的 <a-input-number> 组件 parser 函数.
 *
 * @param value - .
 * @param locale - 国际化字段.
 */
export function parserWeek(value: string, locale: string): number {
    const [key] = Object.entries(I18n[locale][WEEK]).find(([, val]) => val === value);

    return WEEKS.findIndex(({ value }) => value === key);
}


src/shared/constants.ts
export const MIN_SECOND = 0;
export const MAX_SECOND = 59;

export const MIN_MINUTE = 0;
export const MAX_MINUTE = 59;

export const MIN_HOUR = 0;
export const MAX_HOUR = 23;

export const MIN_DATE = 1;
export const MAX_DATE = 31;

export const MIN_MONTH = 1;
export const MAX_MONTH = 12;

export const MIN_WEEK = 0;
export const MAX_WEEK = 6;

export const MIN_YEAR = new Date().getFullYear();
export const MAX_YEAR = 2099;

export const SUNDAY = 'sunday';
export const MONDAY = 'monday';
export const TUESDAY = 'tuesday';
export const WEDNESDAY = 'wednesday';
export const THURSDAY = 'thursday';
export const FRIDAY = 'friday';
export const SATURDAY = 'saturday';

export const WEEKS = [
    { value: SUNDAY, abbr: 'SUN', index: '0' },
    { value: MONDAY, abbr: 'MON', index: '1' },
    { value: TUESDAY, abbr: 'TUE', index: '2' },
    { value: WEDNESDAY, abbr: 'WED', index: '3' },
    { value: THURSDAY, abbr: 'THU', index: '4' },
    { value: FRIDAY, abbr: 'FRI', index: '5' },
    { value: SATURDAY, abbr: 'SAT', index: '6' },
];

export const WEEK_INDEX_REGEXP = new RegExp(WEEKS.map(({ index }) => `(?<!#)${index}`).join('|'), 'g');
export const WEEK_LETTER_REGEXP = new RegExp(WEEKS.map(({ abbr }) => abbr).join('|'), 'g');

export const TYPE = {
    EVERY: '*',
    RANGE: '-',
    STEP: '/',
    SPECIFY: ',',
    UNSPECIFIC: '?',
    EMPTY: '',
    LAST_DAY: 'L',
    LAST_WEEKDAY: 'LW',
    WELL: '#',
    WEEKDAY: 'W',

};

export const SECOND = 'second';
export const MINUTE = 'minute';
export const HOUR = 'hour';
export const DATE = 'date';
export const MONTH = 'month';
export const WEEK = 'week';
export const YEAR = 'year';

export const FIELDS = [
    { value: SECOND, min: MIN_SECOND, max: MAX_SECOND },
    { value: MINUTE, min: MIN_MINUTE, max: MAX_MINUTE },
    { value: HOUR, min: MIN_HOUR, max: MAX_HOUR },
    { value: DATE, min: MIN_DATE, max: MAX_DATE },
    { value: MONTH, min: MIN_MONTH, max: MAX_MONTH },
    { value: WEEK, min: MIN_WEEK, max: MAX_WEEK },
    { value: YEAR, min: MIN_YEAR, max: MAX_YEAR },
];

export const LOCALE_CN = 'cn';
export const LOCALE_EN = 'en';

export const DEFAULT_CRON_EXPRESSION = '* * * * * ?';
export const DEFAULT_LOCALE = LOCALE_CN;

